### PR TITLE
Switch to gimli for DWARF parsing

### PR DESCRIPTION
**Pull request by danielocfb**
Please see individual commits for descriptions.

```
 CHANGELOG.md            |   7 +
 Cargo.toml              |   5 +-
 README.md               |   4 +-
 build.rs                |   6 +-
 data/dwarf-example      | Bin 2099328 -> 0 bytes
 src/dwarf/constants.rs  |  64 -----
 src/dwarf/debug_info.rs | 896 --------------------------------------------------------------------
 src/dwarf/mod.rs        |   4 -
 src/dwarf/parser.rs     | 409 +++++++++++++++----------------
 src/elf/parser.rs       |   7 -
 src/util.rs             |   5 -
 11 files changed, 204 insertions(+), 1203 deletions(-)
```

### CODE REVIEW

File: src/dwarf/parser.rs [hunk 1]
```diff
@@ -684,136 +700,186 @@ fn find_die_sibling(die: &mut debug_info::DIE<'_>) -> Option<usize> {
...
-                            .to_str()
-                            .map_err(|_e| {
-                                Error::new(
-                                    ErrorKind::InvalidData,
-                                    "fail to extract the name of a subprogram",
-                                )
-                            })?
-                    },
-                    debug_info::AttrValue::String(s) => s,
+                let attr_name = attr
```
Comment (anakryiko) [inline src/dwarf/parser.rs:732]: we seem to need this only for error/warning reporting, yet we pay the overhead each time. Can this be done lazily?

Comment (anakryiko) [reply src/dwarf/parser.rs:732]: reading further, seems like exactly what you are doing for DW_AT_low_pc and DW_AT_high_pc, so why not the same here?

File: src/elf/parser.rs [hunk 2]
```diff
@@ -272,12 +273,14 @@ impl<'mmap> Cache<'mmap> {
             return Ok(strtab)
         }
 
-        let idx = if let Ok(idx) = self.find_section(".strtab") {
-            idx
+        let strtab = if let Some(idx) = self.find_section(".strtab")? {
```
Comment (anakryiko) [inline src/elf/parser.rs:276]: this is not something that should be fixed in this PR, but it feels a bit sloppy to have this "if not .strtab, then .dynstr" assumption, because in practice you can have both. E.g., looking at libbpf.so, I see:

```
Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .note.gnu.build-id NOTE            0000000000000238 000238 000024 00   A  0   0  4
  [ 2] .gnu.hash         GNU_HASH        0000000000000260 000260 000a50 00   A  3   0  8
  [ 3] .dynsym           DYNSYM          0000000000000cb0 000cb0 002940 18   A  4   1  8
  [ 4] .dynstr           STRTAB          00000000000035f0 0035f0 001f3b 00   A  0   0  1
  [ 5] .gnu.version      VERSYM          000000000000552c 00552c 000370 02   A  3   0  2
  [ 6] .gnu.version_d    VERDEF          00000000000058a0 0058a0 0002e4 00   A  4  21  8
  [ 7] .gnu.version_r    VERNEED         0000000000005b88 005b88 000130 00   A  4   3  8
  [ 8] .rela.dyn         RELA            0000000000005cb8 005cb8 002b80 18   A  3   0  8
  [ 9] .rela.plt         RELA            0000000000008838 008838 001578 18  AI  3  22  8
  [...]
  [20] .dynamic          DYNAMIC         0000000000057dc8 056dc8 000210 10  WA  4   0  8
  [21] .got              PROGBITS        0000000000057fd8 056fd8 000028 08  WA  0   0  8
  [22] .got.plt          PROGBITS        0000000000058000 057000 000740 08  WA  0   0  8
  [...]
  [34] .symtab           SYMTAB          0000000000000000 1dfc40 004cf8 18     35 382  8
  [35] .strtab           STRTAB          0000000000000000 1e4938 003db2 00      0   0  1
  [36] .shstrtab         STRTAB          0000000000000000 1e86ea 000178 00      0   0  1
```

Note how libbpf.so has both .dynstr and .strtab, and .symtab and .dynsym. And how ("Lk" column") different sections reference different STRTAB sections.

So this might be working on ok right now due to the "use .symtab, otherwise fallback to .dynsym", but it seems like we are painting ourselves into the corner with these assumption.

Basically, ELF section itself declares section index of associated STRTAB section for string references, so ideally we just follow that for whatever ELF section we are working with.

**Review by anakryiko — APPROVED** [review]
Looks like a pretty straightforward switch, great job! I left some comments for follow ups about ELF STRTAB assumptions, but it's definitely not part of this PR.

File: src/dwarf/parser.rs (hunk 1 above)
Comment (danielocfb) [reply src/dwarf/parser.rs:732]: I just didn't want to repeat the same thing three times. But will move that calculation into a function.

File: src/elf/parser.rs (hunk 2 above)
Comment (danielocfb) [reply src/elf/parser.rs:276]: Makes sense. Thanks for pointing that out. We should most definitely use the expected string data section (and probably shouldn't use a heuristic such as "if not .strtab, then .dynstr"). Will take care of that as part of https://github.com/libbpf/blazesym/issues/64.
